server/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetaDataUpdater.java

```diff
@@ -69,6 +69,13 @@ public void shardInitialized(ShardRouting unassignedShard, ShardRouting initiali
     @Override
     public void shardStarted(ShardRouting initializingShard, ShardRouting startedShard) {
         addAllocationId(startedShard);
+        if (startedShard.primary()
+            // started shard has to have null recoverySource; have to pick up recoverySource from its initializing state
+            && (initializingShard.recoverySource() instanceof RecoverySource.ExistingStoreRecoverySource
```

Comment:
Why are snapshot and restore relevant here? I hope we can make this more explicit, depending on `ExistingStoreRecoverySource#FORCE_STALE_PRIMARY_INSTANCE` in some form or fashion

Reply:
fbdf6d7, details here: https://github.com/elastic/elasticsearch/pull/34140#discussion_r222284550